src/main/scala/mesosphere/marathon/state/Group.scala

```diff
@@ -152,32 +153,79 @@ object Group extends StrictLogging {
   def validGroup(base: PathId, enabledFeatures: Set[String]): Validator[Group] =
     validator[Group] { group =>
       group.id is validPathWithBase(base)
-      group.apps.values as "apps" is every(
-        AppDefinition.validNestedAppDefinition(group.id.canonicalPath(base), enabledFeatures))
+
+      group.transitiveApps as "apps" is everyApp(
+        AppDefinition.validBasicAppDefinition(enabledFeatures) and isChildOfParentId(group)
+      )
+
       group is noAppsAndPodsWithSameId
       group is noAppsAndGroupsWithSameName
       group is noPodsAndGroupsWithSameName
-      group.groupsById.values as "groups" is every(validGroup(group.id.canonicalPath(base), enabledFeatures))
+
+      group.transitiveGroupValues as "groups" is every(
+        noAppsAndPodsWithSameId and noAppsAndGroupsWithSameName and noPodsAndGroupsWithSameName
+          and isTrue("Group has to be child of groups with parent id") { childGroup =>
+            if (childGroup.id.parent == group.id) group.groupsById.contains(childGroup.id)
+            else {
+              group.group(childGroup.id.parent) match {
+                case None => false
+                case Some(parentGroup) => parentGroup.groupsById.contains(childGroup.id)
+              }
+            }
+          }
+      )
     }
 
+  implicit def everyApp(validator: Validator[AppDefinition]): Validator[Iterable[AppDefinition]] = {
+    new Validator[Iterable[AppDefinition]] {
+      override def apply(seq: Iterable[AppDefinition]): Result = {
+        seq.foldLeft[Result](Success) {
+          case (accum, (app)) =>
+            validator(app) match {
+              case Success => accum
+              case Failure(violations) =>
+                val scopedViolations = violations.map { violation =>
+                  violation.withPath(Descriptions.Explicit(app.id.toString))
+                }
+                accum.and(Failure(scopedViolations))
+            }
+        }
+      }
+    }
+  }
+
+  private def isChildOfParentId(group: Group): Validator[AppDefinition] = {
+    isTrue("App has to be child of group with parent id") { app =>
+      if (app.id.parent == group.id) group.apps.contains(app.id)
+      else {
+        group.group(app.id.parent) match {
+          case None => false
+          case Some(parentGroup) => parentGroup.apps.contains(app.id)
```

Comment:
Wdyt about just checking the ids e.g. smth. like `group.group(app.id.parent).id == group.id`?

Reply:
...or just group equality: `group.group(app.id.parent) == group` since `equals` already checking everything?

Reply:
I'm not sure this is the same. Basically we want to avoid this case

```json
{
    "id": "/my_group",
    "apps": { "/different_group/my_app": {"cmd": "sleep"}}
}
{
    "id":"/different_group",
    "apps": {"/different_group/another_app": {"cmd":"sleep"}}
}
```
So for app `my_app` we fetch it's parent `different_group` and check that it does actually hold the app. As I've mentioned we probably should do this check during the creation of a group or an `updateApp` and `updatePod` call.

Reply:
Ah, I get it now. `parentGroup` confused me a bit - I'd rename it to `childGroup`